### ISSUE TITLE
FetchUser/Middleware Race Condition

### ISSUE THREAD

**Issue by sschadwick** [nuxt-community/auth-module#169]
### Version

[v4.3.0](https://github.com/nuxt-community/auth-module/releases/tag/v4.3.0)

### Steps to reproduce

Currently running nuxt-js via express middleware on an AWS Lambda, and using AWS Cognito as an OAUTH2 authorizer for a SAML IdP. 

### What is expected ?

It is expected that an authenticated user is able to directly navigate to nested pages in the app via deep links.

Additionally, if an unauthenticated user navigates to a deep link, they should be redirected to that link after they have authenticated.

### What is actually happening?

When directly navigating to a deep link (or refreshing any page other than `/`), the app will redirect the user to `/login` and subsequently `/`. Interestingly, this behavior rarely occurs when developing locally, however it appears quite frequently when deployed to stage/prod. As a result, an authenticated user is never able to load specific pages unless they navigate to them within the web app.

### Additional comments?

I believe that there is a race condition involving the auth middleware and the strategy-specific `fetchUser` method. Although the oauth2 `fetchUser` method is async, the `fetchUser` method in `core/auth.js` does not await upon the result. This explains why the observed behavior is inconsistent.

Furthermore, I added comments to `core/auth.js`, `core/storage`, and `core/middleware` in order to troubleshoot why the middleware was evaluating `ctx.app.$auth.$state.loggedIn` as false. I've attached a picture of my AWS logs, with the corresponding user behavior.

Of note is that when the user follows a deep link or refreshes, the GET request to `userinfo_endpoint` does not resolve before the middleware decides to redirect the user to `/login`.
![debugging nuxt-auth.png](https://cmty.s3.eu-west-1.amazonaws.com/102371998/ebfedd10-52ef-11e8-a403-df62268c86d8.png)

<div align="right"><sub><em>This bug report is available on <a href="https://nuxtjs.cmty.io">Nuxt.js</a> community (<a href="https://nuxtjs.cmty.io/nuxt-community/auth-module/issues/c127">#c127</a>)</em></sub></div>

**Comment by pedantic-git** [comment]
I think I'm experiencing this same issue in a pure Nuxt app using the local strategy - reloading a page with the middleware enabled causes the same kind of race condition and a redirect to the / page.

**Comment by sschadwick** [comment]
I've solved the issue here but unfortunately it was specific to my environment so I'm closing the issue. Once I logged the `.catch` on the ajax request to `userinfo_endpoint`, I found that it was regularly returning a 413.

The issue in my case was that API Gateway I was using for `userinfo_endpoint` (Cognito doesn't provide a native endpoitn) was erroring out because I was exceeding the maximum HTTP Header Size allowed. Although my token is typically under 2000 bytes and the maximum allowed is 10240, AWS copies the entire request header into a `x-apigateway-event` header. The auth token along with non-auth cookies was pushing the count above the limit, causing the user lookup to fail.

Note: There is a lingering bug somewhere that prevents errors from propagating when the ajax request to `userinfo_endpoint` fails but I was unable to find it.
